src/nccl_ofi_net.c

```diff
@@ -158,85 +158,204 @@ void nccl_net_ofi_free_info_list(struct fi_info *info_list)
- * no key was available.
+ * Initializes a nccl_ofi_id_pool_t object, marking all IDs as available.
  *
- * This operation is locked by the key pool's internal lock.
+ * @param	id_pool
+ *		The ID pool
+ * @param	size
+ *		Size of the id pool (number of IDs)
+ * @return	0 on success
+ *		non-zero on error
+ */
+int nccl_ofi_id_pool_init(nccl_ofi_id_pool_t *id_pool, size_t size)
+{
+	int ret = 0;
+
+	if (OFI_UNLIKELY(NULL == id_pool)) {
+		NCCL_OFI_WARN("Invalid call to nccl_ofi_id_pool_init with null pointer");
+		return -EINVAL;
+	}
+
+	/* Scale pool size to number of 64-bit entries (rounded up) */
+	id_pool->entries = (size + sizeof(uint64_t) * 8 - 1) / (sizeof(uint64_t) * 8);
+
+	/* Allocate memory for the pool */
+	id_pool->ids = malloc(sizeof(uint64_t) * id_pool->entries);
+
+	/* Return in case of allocation error */
+	if (OFI_UNLIKELY(NULL == id_pool->ids)) {
+		NCCL_OFI_WARN("Unable to allocate ID pool");
+		return -ENOMEM;
+	}
+
+	/* Set all IDs to be available */
+	memset(id_pool->ids, 0xff, size / 8);
+	if (size % 8) {
+		id_pool->ids[id_pool->entries - 1] = (1ULL << (size % (sizeof(uint64_t) * 8))) - 1;
+	}
+
+	/* Initialize mutex */
+	ret = pthread_mutex_init(&id_pool->lock, NULL);
+	if (OFI_UNLIKELY(ret)) {
+		NCCL_OFI_WARN("Unable to initialize mutex");
+		free(id_pool->ids);
+		id_pool->ids = NULL;
+		return ret;
+	}
+
+	id_pool->size = size;
+
+	return ret;
+}
+
+/*
+ * @brief	Allocate an ID
+ *
+ * Extract an available ID from the ID pool, mark the ID as
+ * unavailable in the pool, and return extracted ID. No-op in case
+ * no ID was available.
+ *
+ * This operation is locked by the ID pool's internal lock.
  *
- * @param	key_pool
- *		The Key pool
- * @return	Extracted key, on susccess
- *		FI_KEY_NOTAVAIL, in case no key is available
+ * @param	id_pool
+ *		The ID pool
+ * @param	id
+ *		Set to the extracted ID (zero-based), on success
+ * @return	0 on success
+ *		non-zero on error
  */
-uint64_t nccl_net_ofi_allocate_mr_key(nccl_ofi_mr_keypool_t *key_pool)
+int nccl_net_ofi_allocate_id(nccl_ofi_id_pool_t *id_pool, uint64_t *id)
 {
-	uint64_t key = FI_KEY_NOTAVAIL;
-	bool* mr_keys = key_pool->mr_keys;
-	int num_mr_keys = key_pool->size;
-	pthread_mutex_t *lock = &key_pool->lock;
+	if (OFI_UNLIKELY(NULL == id_pool || NULL == id)) {
+		NCCL_OFI_WARN("Invalid call to nccl_net_ofi_allocate_id with null pointer");
+		return -EINVAL;
+	}
 
-	if (mr_keys == NULL) {
-		NCCL_OFI_WARN("Invalid call to allocate_mr_key");
-		return FI_KEY_NOTAVAIL;
+	if (OFI_UNLIKELY(NULL == id_pool->ids)) {
+		NCCL_OFI_WARN("Invalid call to nccl_net_ofi_allocate_id with uninitialized pool");
+		return -EINVAL;
 	}
 
-	pthread_mutex_lock(lock);
+	pthread_mutex_lock(&id_pool->lock);
 
-	for (size_t i = 0; i < num_mr_keys; i++) {
-		if (mr_keys[i]) {
-			mr_keys[i] = false;
-			key = i;
+	int entry_index = 0;
+	for (size_t i = 0; i < id_pool->entries; i++) {
+		entry_index = __builtin_ffsll(id_pool->ids[i]);
```

Comment:
You should have a configure test to make sure __builtin_ffsll() works with the compiler and abort if not.